Add solution for CSES Raab Game I (#5770)
cpinitiative/usaco-guide: solutions/bronze/cses-3399.mdx

---
id: cses-3399
source: CSES
title: Raab Game I
author: Zayd Krunz
---

<Spoiler title="Hint 1">
First, consider the draws. In a draw, nobody wins. If $a+b$ is the number of games that are won, how many games must be draws?
</Spoiler>

<Spoiler title="Hint 2">
After handling the draws, we are left with a subproblem where the total number of cards is $k = a+b$.
Since both players have the exact same set of cards $\{1, 2, \dots, k\}$, the sum of their card values is equal.
Is it possible for one player to win *every* round (i.e., score $k$) if the sets are identical?
</Spoiler>

<Spoiler title="Hint 3">
Try a cyclic shift. If Player 1 plays cards in increasing order $1, 2, \dots, k$, try having Player 2 play cards shifted by $a$ (modulo $k$). For example, if $k=5$, Player 1 might play $1, 2, 3, 4, 5$, and Player 2 will play $2, 3, 4, 5, 1$.
</Spoiler>

<Spoiler title="Solution">

## Explanation

Let $N$ be the number of cards and $a$ and $b$ be the target scores.

> First, we handle the draws. The number of draws required is $D = N - (a + b)$. If $D < 0$, the result is impossible (as the total games played must sum to $N$). Otherwise, we can satisfy these draws by having both players play their largest $D$ cards against each other (e.g., $N$ vs $N$, $N-1$ vs $N-1$, etc.).

Reviewer: ```suggestion
First, we handle the draws. The number of draws required is $D = N - (a + b)$. If $D < 0$, the result is impossible since the games played must sum to $N$. Otherwise, we can satisfy these draws by having both players play their largest $D$ cards against each other (e.g., $N$ vs $N$, $N-1$ vs $N-1$, etc.).
```
